docs/proposals/proxy-extensions.md

```diff
@@ -0,0 +1,294 @@
+CRD should be enhanced to provide the ability to define backend services
+that will be used by the extension. Once configured UI can send requests
+to API server in a specific endpoint. API server will act as a reverse
+proxy receiving the request from the UI and routing to the appropriate
+backend service.
+
+Example:
+```yaml 
+apiVersion: argoproj.io/v1alpha1
+kind: ArgoCDExtension
+metadata:
+  name: my-cool-extention
+  finalizers:
+    - extensions-finalizer.argocd.argoproj.io
+spec:
+  sources:
+    - git:
+        url: https://github.com/some-org/my-cool-extension.git
+  backend:
+    serviceName: some-backend-svc
+    endpoint: /some-backend
+```
+
+**Note**: While this is a nice-to-have, it won't be part of the first proxy
+extension version. This would need to be considered if ArgoCD extensions
+eventually get traction.
+
+## Non-Goals
+
+It isn't in the scope of this proposal to specify commands in the ArgoCD
+CLI. This proposal covers the reverse-proxy extension spec that will be
+used by ArgoCD UI.
+
+## Proposal
+
+### Use cases
+
+The following use cases should be implemented for the conclusion of this
+proposal:
+
+#### [UC-1]: As an ArgoCD admin, I want to configure a backend services so it can be used by my UI extension
+
+Define a new section in the ArgoCD configmap ([argocd-cm.yaml][4])
+allowing admins to register and configure new extensions. All enabled
+extensions backend will be available to be invoked by the ArgoCD UI under
+the following API base path:
+
+`<argocd-host>/api/v1/extensions/<extension-name>`
+
+
+With the configuration bellow, the expected behaviour is explained in the
+following examples:
+
+```yaml
+extension.config:
+| extensions:
+    - name: some-extension
+      enabled: true
+      backend:
+        idleConnTimeout: 10s
+        url: http://extension-host.com:8080
+```
+
+- **Example 1**:
+
+ArgoCD API server acts as a reverse-proxy forwarding http requests as
+follows:
+
+```
+
+   ┌────────────┐
+   │ ArgoCD UI  │
+   └──────┬─────┘
+          │
+          │ GET http://argo.com/api/v1/extensions/some-extension
+          │
+          ▼
+ ┌─────────────────┐
+ │ArgoCD API Server│
+ └────────┬────────┘
+          │
+          │ GET http://extension-host.com:8080
+          │
+          ▼
+  ┌───────────────┐
+  │Backend Service│
+  └───────────────┘
+
+```
+
+- **Example 2**:
+
+If a backend provides an API under the `/apiv1/metrics` endpoint, ArgoCD
+should be able to invoke it such as:
+
+```
+   ┌────────────┐
+   │ ArgoCD UI  │
+   └──────┬─────┘
+          │
+          │ GET http://argo.com/api/v1/extensions/some-extension/apiv1/metrics/123
+          │
+          ▼
+ ┌─────────────────┐
+ │ArgoCD API Server│
+ └────────┬────────┘
+          │
+          │ GET http://extension-host.com:8080/apiv1/metrics/123
+          │
+          ▼
+  ┌───────────────┐
+  │Backend Service│
+  └───────────────┘
+```
+
+##### Considerations
```

Comment:
investigate kubernetes aggregate api implementation to check if we can reuse something:
Are we drawing any inspiration how k8s does extension via aggregate apiserver?https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#register-apiservice-objects
Are we drawing any inspiration how k8s does extension via aggregate apiserver?https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#register-apiservice-objects

Reply:
We could implement a security layer between Argo CD API Server and the backend service similar to how kubernetes aggregate API Server is designed to authenticate with extensions. We could use a self-signed x509 certificate to the request that originates in Argo CD API server targeting the backend service. The backend service could then use the certificate to ensure the request is really coming from Argo CD API server.

Reply:
Added goal 6